src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs

```diff
@@ -302,174 +305,99 @@ public class OpenIdConnectAuthenticationHandler : AuthenticationHandler<OpenIdCo
                     _configuration = await Options.ConfigurationManager.GetConfigurationAsync(Context.RequestAborted);
                 }
 
-                AuthenticationTicket ticket = null;
-                JwtSecurityToken jwt = null;
-
-                // OpenIdConnect protocol allows a Code to be received without the id_token
-                if (!string.IsNullOrEmpty(message.IdToken))
+                if (string.IsNullOrEmpty(message.IdToken) && !string.IsNullOrEmpty(message.Code))
```

Comment:
This is getting complicated enough that I recommend extracting it to its own method. Same goes for the else case.